### PR TITLE
[WIP] New Backend "GeoSQL Backend for PowerDNS"

### PR DESCRIPTION
**Pull request by shinsterneck**
I would like to contribute the GeoSQL backend for PowerDNS. It enables geo dns records to be stored inside an SQL database and is meant to simplify creation and administration of geoip enabled zones.

Please have a look at the readme file in the backend source directory for more information. (modules/geosqlbackend/README)

I hope this gets accepted and I am willing to keep maintaining and improving it regularly.

Best Regards,
Shin

### CODE REVIEW

**Comment by cmouse** [comment]
regression test suite would be nice.

**Comment by shinsterneck** [comment]
Thanks for your quick response. I'll have a look at the other tests and try to create one for this backend.

**Comment by ahupowerdns** [comment]
Thank you! We're studying your work, it looks very interesting.  Bert.

**Comment by pieterlexis** [comment]
@shinsterneck Do let us know when you think this PR is finished so we can take a look at it :+1:

**Comment by shinsterneck** [comment]
@pieterlexis Thanks! I am having a bit of trouble understanding how I can integrate the backend specific regression tests so they get called. Can you give me some pointers on this?

**Comment by cmouse** [comment]
Look at how remotebackend does these

**Comment by cmouse** [comment]
Or geoipbackend

**Comment by shinsterneck** [comment]
I just looked at the start-test-stop script and found the "start_master" function being called there, so I think I understood now how this works. I will work on this tomorrow as its already 1:20am here in Japan ;-)

Thanks!

**Comment by pieterlexis** [comment]
@shinsterneck  Could you rebase this and let us know if you run into any issues with the cration of the tests?

**Comment by shinsterneck** [comment]
will do, please give me until tomorrow as I am traveling now.

**Comment by shinsterneck** [comment]
ok I rebased the branch now, but getting travis issues. looking into them now...

**Comment by pieterlexis** [comment]
Interesting failure on travis,
```
checking for the Boost regex library... configure: error: cannot compile a test that uses Boost regex
```

The `config.log` I got from a Ubuntu 12.04 VM (which also fails) tells me:
```
In file included from /usr/include/boost/shared_ptr.hpp:17:0,
                 from /usr/include/boost/regex/v4/regex_workaround.hpp:43,
                 from /usr/include/boost/regex/v4/regex.hpp:32,
                 from /usr/include/boost/regex.hpp:31,
                 from conftest.cpp:39:
/usr/include/boost/smart_ptr/shared_ptr.hpp: In instantiation of 'boost::shared_ptr<T>& boost::shared_ptr<T>::operator=(const boost::shared_ptr<T>&) [with T = boost::re_detail::named_subexpressions]':
/usr/include/boost/regex/v4/match_results.hpp:549:20:   required from 'void boost::match_results<BidiIterator, Allocator>::set_named_subs(boost::shared_ptr<boost::re_detail::named_subexpressions>) [with BidiIterator = const char*; Allocator = std::allocator<boost::sub_match<const char*> >]'
/usr/include/boost/regex/v4/perl_matcher_common.hpp:203:4:   required from 'bool boost::re_detail::perl_matcher<BidiIterator, Allocator, traits>::match_imp() [with BidiIterator = const char*; Allocator = std::allocator<boost::sub_match<const char*> >; traits = boost::regex_traits<char>]'
/usr/include/boost/regex/v4/perl_matcher_common.hpp:180:21:   required from 'bool boost::re_detail::perl_matcher<BidiIterator, Allocator, traits>::match() [with BidiIterator = const char*; Allocator = std::allocator<boost::sub_match<const char*> >; traits = boost::regex_traits<char>]'
/usr/include/boost/regex/v4/regex_match.hpp:50:25:   required from 'bool boost::regex_match(BidiIterator, BidiIterator, boost::match_results<Iterator, Allocator>&, const boost::basic_regex<charT, traits>&, boost::regex_constants::match_flag_type) [with BidiIterator = const char*; Allocator = std::allocator<boost::sub_match<const char*> >; charT = char; traits = boost::regex_traits<char>; boost::regex_constants::match_flag_type = boost::regex_constants::_match_flags]'
/usr/include/boost/regex/v4/regex_match.hpp:90:95:   required from 'bool boost::regex_match(const charT*, const boost::basic_regex<charT, traits>&, boost::regex_constants::match_flag_type) [with charT = char; traits = boost::regex_traits<char>; boost::regex_constants::match_flag_type = boost::regex_constants::_match_flags]'
conftest.cpp:44:53:   required from here
/usr/include/boost/smart_ptr/shared_ptr.hpp:305:9: error: use of deleted function 'boost::shared_ptr<boost::re_detail::named_subexpressions>::shared_ptr(const boost::shared_ptr<boost::re_detail::named_subexpressions>&)'
         this_type(r).swap(*this);
         ^
/usr/include/boost/smart_ptr/shared_ptr.hpp:168:25: note: 'boost::shared_ptr<boost::re_detail::named_subexpressions>::shared_ptr(const boost::shared_ptr<boost::re_detail::named_subexpressions>&)' is implicitly declared as deleted because 'boost::shared_ptr<boost::re_detail::named_subexpressions>' declares a move constructor or move assignment operator
 template<class T> class shared_ptr
                         ^
In file included from /usr/include/boost/regex/v4/regex.hpp:82:0,
                 from /usr/include/boost/regex.hpp:31,
                 from conftest.cpp:39:
/usr/include/boost/regex/v4/match_results.hpp: In instantiation of 'boost::match_results<BidiIterator, Allocator>::match_results(const boost::match_results<BidiIterator, Allocator>&) [with BidiIterator = const char*; Allocator = std::allocator<boost::sub_match<const char*> >]':
/usr/include/boost/regex/v4/perl_matcher_non_recursive.hpp:134:76:   required from 'boost::re_detail::saved_recursion<Results>::saved_recursion(int, const boost::re_detail::re_syntax_base*, Results*) [with Results = boost::match_results<const char*>]'
/usr/include/boost/regex/v4/perl_matcher_non_recursive.hpp:342:4:   required from 'void boost::re_detail::perl_matcher<BidiIterator, Allocator, traits>::push_recursion(int, const boost::re_detail::re_syntax_base*, boost::re_detail::perl_matcher<BidiIterator, Allocator, traits>::results_type*) [with BidiIterator = const char*; Allocator = std::allocator<boost::sub_match<const char*> >; traits = boost::regex_traits<char>; boost::re_detail::perl_matcher<BidiIterator, Allocator, traits>::results_type = boost::match_results<const char*>]'
/usr/include/boost/regex/v4/perl_matcher_non_recursive.hpp:935:127:   required from 'bool boost::re_detail::perl_matcher<BidiIterator, Allocator, traits>::match_endmark() [with BidiIterator = const char*; Allocator = std::allocator<boost::sub_match<const char*> >; traits = boost::regex_traits<char>]'
/usr/include/boost/regex/v4/perl_matcher_non_recursive.hpp:147:7:   required from 'bool boost::re_detail::perl_matcher<BidiIterator, Allocator, traits>::match_all_states() [with BidiIterator = const char*; Allocator = std::allocator<boost::sub_match<const char*> >; traits = boost::regex_traits<char>]'
/usr/include/boost/regex/v4/perl_matcher_common.hpp:323:21:   required from 'bool boost::re_detail::perl_matcher<BidiIterator, Allocator, traits>::match_prefix() [with BidiIterator = const char*; Allocator = std::allocator<boost::sub_match<const char*> >; traits = boost::regex_traits<char>]'
/usr/include/boost/regex/v4/perl_matcher_common.hpp:207:25:   required from 'bool boost::re_detail::perl_matcher<BidiIterator, Allocator, traits>::match_imp() [with BidiIterator = const char*; Allocator = std::allocator<boost::sub_match<const char*> >; traits = boost::regex_traits<char>]'
/usr/include/boost/regex/v4/perl_matcher_common.hpp:180:21:   required from 'bool boost::re_detail::perl_matcher<BidiIterator, Allocator, traits>::match() [with BidiIterator = const char*; Allocator = std::allocator<boost::sub_match<const char*> >; traits = boost::regex_traits<char>]'
/usr/include/boost/regex/v4/regex_match.hpp:50:25:   required from 'bool boost::regex_match(BidiIterator, BidiIterator, boost::match_results<Iterator, Allocator>&, const boost::basic_regex<charT, traits>&, boost::regex_constants::match_flag_type) [with BidiIterator = const char*; Allocator = std::allocator<boost::sub_match<const char*> >; charT = char; traits = boost::regex_traits<char>; boost::regex_constants::match_flag_type = boost::regex_constants::_match_flags]'
/usr/include/boost/regex/v4/regex_match.hpp:90:95:   required from 'bool boost::regex_match(const charT*, const boost::basic_regex<charT, traits>&, boost::regex_constants::match_flag_type) [with charT = char; traits = boost::regex_traits<char>; boost::regex_constants::match_flag_type = boost::regex_constants::_match_flags]'
conftest.cpp:44:53:   required from here
/usr/include/boost/regex/v4/match_results.hpp:81:130: error: use of deleted function 'boost::shared_ptr<boost::re_detail::named_subexpressions>::shared_ptr(const boost::shared_ptr<boost::re_detail::named_subexpressions>&)'
       : m_subs(m.m_subs), m_named_subs(m.m_named_subs), m_last_closed_paren(m.m_last_closed_paren), m_is_singular(m.m_is_singular) 
                                                                                                                                  ^
```

The failed program was:

```
/* confdefs.h */
#define PACKAGE_NAME "pdns"
#define PACKAGE_TARNAME "pdns"
#define PACKAGE_VERSION "0.0.HEAD.g"
#define PACKAGE_STRING "pdns 0.0.HEAD.g"
#define PACKAGE_BUGREPORT ""
#define PACKAGE_URL ""
#define PDNS_CONFIG_ARGS " '--with-modules=geosql'"
#define PACKAGE "pdns"
#define VERSION "0.0.HEAD.g"
#define YYTEXT_POINTER 1
#define _GNU_SOURCE 1
#define STDC_HEADERS 1
#define HAVE_SYS_TYPES_H 1
#define HAVE_SYS_STAT_H 1
#define HAVE_STDLIB_H 1
#define HAVE_STRING_H 1
#define HAVE_MEMORY_H 1
#define HAVE_STRINGS_H 1
#define HAVE_INTTYPES_H 1
#define HAVE_STDINT_H 1
#define HAVE_UNISTD_H 1
#define HAVE_DLFCN_H 1
#define LT_OBJDIR ".libs/"
#define HAVE_TM_GMTOFF 1
#define HAVE_SYS_MMAN_H 1
#define HAVE_MMAP 1
#define HAVE_BOOST 1
#define HAVE_BOOST_FOREACH_HPP 1
#define HAVE_BOOST_PROGRAM_OPTIONS_HPP 1
#define HAVE_BOOST_ARCHIVE_TEXT_OARCHIVE_HPP 1
#define BUILD_HOST "root@6dcb3a5c911a"
#define HAVE_STRCASESTR 1
#define HAVE_LOCALTIME_R 1
#define HAVE_RECVMMSG 1
#define PDNS_MODULES "geosql"
#define HAVE_BOOST_REGEX_HPP 1
/* end confdefs.h.  */
#include <boost/regex.hpp>

int
main ()
{
boost::regex exp("*"); boost::regex_match("foo", exp);
  ;
  return 0;
}
```

I have a feeling that the test we use for Boost::Regex requires a newer Boost than the 1.47 from Ubuntu 12.04. I'm currently compiling Boost 1.58 on Ubuntu 12.04 to test (to be continued)

**Comment by pieterlexis** [comment]
This problem indeed goes away with a newer boost. If you apply the following (untested) patch, Travis should be happy again:

```
diff --git a/.travis.yml b/.travis.yml
index 602fae5..7ea0bba 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -11,6 +11,7 @@ before_script:
  - sudo apt-get update -qq
  - sudo apt-get -qq install g++-4.8
  - sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-4.8 90
+ - sudo add-apt-repository -y ppa:boost-latest/ppa
  - sudo apt-get install --quiet --quiet --no-install-recommends
      alien
      authbind
@@ -21,7 +22,7 @@ before_script:
      faketime
      geoip-database
      ldnsutils
-     libboost-all-dev
+     boost1.55
      libbotan1.10-dev
      libcdb-dev
      libcdb-dev
```

**Comment by shinsterneck** [comment]
@pieterlexis Thanks for investigating this and for the patch!
Yes you are right, the regex test with boost version 1.55 works fine. The only thing I had to change in your patch was to move the line with "ppa:boost-latest/ppa" before the apt-update command so that the package is indexed and apt-get install can find it.

Travis is now running on my development branch and once I confirmed that it works I'll merge the change into this branch.

**Comment by shinsterneck** [comment]
I noticed that the travis checks fails if it takes longer than 50 minutes. Maybe we should split the tests into two seperate processes, one for testing the build process and another for the regression-tests?

**Comment by pieterlexis** [comment]
@shinsterneck The problem with that is that we need the build artifacts to test against :). We're in the process of creating our own CI infra to test this (but it'll take a while). For now , the idea is to run the tests in parallel instead of sequentially.

**Comment by shinsterneck** [comment]
Note that I squashed the commits to make the whole thing more transparent.

**Comment by shinsterneck** [comment]
@pieterlexis Can you help me with this? How would I go about changing these tests to run parallel? Is this something I can do outside of this pull request or would that change need to be coordinated with you?

**Comment by pieterlexis** [comment]
Hi Shin,

I'm currently looking at this, but I have a compile issue of this backend after rebasing on master (might be GCC 5.2 related or a lack of understanding):
```
geosqlbackend.cpp: In member function ‘virtual DNSBackend* GeoSqlFactory::make(const string&)’:
geosqlbackend.cpp:310:40: error: invalid new-expression of abstract class type ‘GeoSqlBackend’
         return new GeoSqlBackend(suffix);
                                        ^
In file included from geosqlbackend.cpp:23:0:
geosqlbackend.h:41:7: note:   because the following virtual functions are pure within ‘GeoSqlBackend’:
 class GeoSqlBackend : public DNSBackend {
       ^
In file included from geosqlbackend.h:32:0,
                 from geosqlbackend.cpp:23:
../../pdns/dnsbackend.hh:114:16: note: 	virtual void DNSBackend::lookup(const QType&, const DNSName&, DNSPacket*, int)
   virtual void lookup(const QType &qtype, const DNSName &qdomain, DNSPacket *pkt_p=0, int zoneId=-1)=0; 
                ^
../../pdns/dnsbackend.hh:122:16: note: 	virtual bool DNSBackend::list(const DNSName&, int, bool)
   virtual bool list(const DNSName &target, int domain_id, bool include_disabled=false)=0;

```

I'm going to setup an Ubuntu 12.04 VM to mimick the travis environment  and try again.

**Comment by cmouse** [comment]
It's missing lookup and list with correct signatures. Even if list is not supported you need to return false.

**Comment by pieterlexis** [comment]
Right, it's because I rebased on master which has DNSName instead of strings. It must be Friday

**Comment by pieterlexis** [comment]
It looks like gcc is killed on travis during `make check`:

```
g++: internal compiler error: Killed (program cc1plus)
```

Compiling with clang works fine. On my local machine (Arch Lnux, GCC 5.2, Boost 1.58) `make check` works (but fails with an access violation in unrelated code). Time to recreate this bug in an environment similar to travis.

**Comment by shinsterneck** [comment]
Yeah I noticed that the master branch has new/different internet definitions, so I'll update my code to relfect those changes.

**Comment by shinsterneck** [comment]
Rebased and Squashed!

**Comment by shinsterneck** [comment]
Rebased and Squashed!

**Comment by shinsterneck** [comment]
@pieterlexis Any ideas why the gcc job is failing while the clang job is a success?

No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.
The build has been terminated
(Reading database ... 76783 files and directories currently installed.)(Reading database ... 76770 files and directories currently installed.)(Reading database ... 76770 files and directories currently installed.)(Reading database ... 76763 files and directories currently installed.)

**Comment by cmouse** [comment]
```

g++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-4.8/README.Bugs> for instructions.
make[2]: *** [dnsdist-tcp.o] Error 4
  CXX    dnsdist-web.o
g++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-4.8/README.Bugs> for instructions.
make[2]: *** [dnsdist.o] Error 4
  CXX    dnslabeltext.o
g++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-4.8/README.Bugs> for instructions.
make[2]: *** [dnsdist-lua.o] Error 4
  CXX    dnsname.o
  CXX    dnswriter.o
  CXX    iputils.o
g++: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-4.8/README.Bugs> for instructions.
make[2]: *** [dnsdist-web.o] Error 4
```

**Comment by shinsterneck** [comment]
Maybe an out of memory issue with Travis.

**Comment by shinsterneck** [comment]
ok looks like the error is gone....

**Comment by shinsterneck** [comment]
FYI: I will rebase this and complete this with the requested tests.

**Comment by TGM** [comment]
Any update?

**Comment by cmouse** [comment]
Will this break DNSSEC for gmysql? Also toLower is not needed with mysql.

**Comment by cmouse** [comment]
Also other places use cc and hh instead of cpp and hpp. It would make sense to conform these. There is also smysql driver you could use for mysql instead of opendbx.

**Comment by Habbie** [comment]
Why would this break DNSSEC for gmysql? I don't see it touching gmysql source at all.

**Comment by Habbie** [comment]
re toLower, you shouldn't count on your schema being configured such that it is not necessary; however, it is still not necessary because pdns core always does it.

**Comment by cmouse** [comment]
re DNSSEC. This claims to be overlay so it does somehow interact with gmysql. Imo it is reasonable to consider that it might also break dnssec since it usually does that in multi-backend setup.

**Comment by Habbie** [comment]
@cmouse I see. Perhaps it gets away with it because gmysql (launched first I assume) holds all the key material?

**Comment by shinsterneck** [comment]
thanks for the feeedback! 

1. I renamed the extension

2. smysql instead of OpenDBX: Yes I have also been thinking about using something else more up to date and more "direct". I used OpenDBX mainly to be able to support other backends, but so far this has not been neccessary and I am relying anyway on the gmysql database to retrieve the main records... --> I'll look into smysql. 

3. Yes looks like I can get rid of toLower.

3. Regarding DNSSEC, yes I think it will break it so I need to think of a way to fix this. Right now the geosql backend would come before the gmysql backend as, at the time when I started, I wasn't able to  inform pdns that the specific records geosql would find are of higher priority then the ones found by other backends before it, so I placed it in front the gmysql backend.  --> Is there a better way?

**Comment by cmouse** [comment]
A solution might be to use cname indirection so geosql could be on a different zone. Would still need to build dnssec support though. If you want, have a look how geoipbackend works. It might give you ideas.

**Comment by shinsterneck** [comment]
Is smysql a PowerDNS made driver or where can I find more on this?

**Comment by cmouse** [comment]
Look under modules/gmysqlbackend/smysql.hh

**Comment by shinsterneck** [comment]
removed the opendbx dependency and use smysql driver instead. Still needs further work but step one is done.

**Comment by cmouse** [comment]
Just a thought. There is sql statement support in ssql. So you can do prepared statements instead of boost::replace_all.

**Comment by shinsterneck** [comment]
@cmouse , yes I have seen this function and will try to convert to it. Thank you!

**Comment by m1911** [comment]
Can this backend be used properly?

**Comment by shinsterneck** [comment]
@m1911 , yes but its still missing the DNSSEC support feature. I unfortunately did not have enough time to get around implementing this, but I will restart my work on this as I will also need it soon ;-) Other than that it works!

**Comment by Habbie** [comment]
> Maybe we should just merge this for now?

Undocumented (as far as doc.powerdns.com goes) and thus invisible, not touching any other files in a dangerous way.. This would work for me.

**Comment by cmouse** [comment]
Documents should be placed into `docs/backends/geosql.rst` and you probably need to update `index.rst` in the same folder

File: modules/geosqlbackend/README.md [hunk 1]
```diff
@@ -0,0 +1,598 @@
...
+of the License, or (at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
```
Comment (zeha) [inline modules/geosqlbackend/README.md]: this should at the very least be gpl2+openssl exception, as all the other powerdns parts

File: modules/geosqlbackend/geosqlbackend.cc [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+                } catch ( std::exception &e ) {
+                    cache_mutex.unlock();
+                    throw PDNSException ( "geosql Error while retrieving region records from the database: " );
+                }
+
+                break;
+            }
+    }
+
+    \
```
Comment (zeha) [inline modules/geosqlbackend/geosqlbackend.cc]: stray \\ ?

**Comment by Habbie** [comment]
If you add some testing (including Travis setup), we will merge this.

**Comment by shinsterneck** [comment]
@Habbie ok will do! Thank you!

**Comment by ahupowerdns** [comment]
So... I know this is not fun. But do you think GeoSQL still has a usecase now that we have https://doc.powerdns.com/authoritative/lua-records.html LUA records, that get us similar functionality for all backends? We (powerdns) are pondering if we want to support both and leaning towards no. And again, I know this is no fun since you've worked on this for so long now..

**Comment by shinsterneck** [comment]
@ahupowerdns I had a look at the LUA record link and think that is a real great feature. I definitely agree with you that it does not make sense to support both methods especially when LUA benefits all backends. I created GeoSQL only because at the time there was no SQL based GeoIP solution available and I wanted to introduce PowerDNS to my colleagues where we can keep everything under "one hood = SQL" and in a simple way (in terms of operations). The LUA records does requires a level of code but I think that the benefits in this case outweigh the requirement of simplicity.

I appreciate your candidness and I am all for the better solution, which in this case I also think is the LUA records feature. Sure its a pitty that my code never made it to the official release, but it was fun to work on it and experience the teams support!

I have the GeoSQL backend in production and will work on testing and moving towards LUA as well. If you are interested I'll share my experiences.

**Comment by ahupowerdns** [comment]
@shinsterneck I'm impressed you take this view! I hope the Lua records are good enough and will not disappoint you. We are very interested in your experiences. If you find anything that is missing, or is not good enough, please let us know so we can be sure the Lua records are a superset of GeoSQL. Thanks!

**Comment by bruciebruce** [comment]
Is this project active?  I am looking to install a PowerDNS cluster but without having to do a yaml file version of geoIP.  Thanks.

**Comment by Habbie** [comment]
> Is this project active? I am looking to install a PowerDNS cluster but without having to do a yaml file version of geoIP. Thanks.

Please read the previous comments about the LUA record.

**Comment by bruciebruce** [comment]
This makes no sense to me.  Having data in a database allows other people in organizations to data mine what other people in the company are doing.  How is someone supposed to use yaml files and discover semantic relationships if this is not what they do full time?

**Comment by bruciebruce** [comment]
I am confused about this.  What does LUA scripting have to do with using a database to store records instead of in a YAML file?

**Comment by pieterlexis** [comment]
> I am confused about this. What does LUA scripting have to do with using a database to store records in instead of in a YAML file? I would like to use a geo database that is accessible (by say the marketing department), so they can see regions of global coverage for strategy planning and data analytics.

see [this](https://blog.powerdns.com/2017/12/15/powerdns-authoritative-lua-records/) and [this](https://doc.powerdns.com/authoritative/lua-records/index.html)

**Comment by pieterlexis** [comment]
Closing, as the LUA record will be the future

**Pull request closed without merging**